device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/auth/IotHubSasTokenAuthenticationProvider.java

```diff
@@ -76,10 +76,13 @@ public void setTokenValidSecs(long tokenValidSecs)
     }
 
     /**
-     * Returns true if the saved sas token needs to be manually renewed by the user
-     * @return true if the saved sas token needs to be manually renewed by the user
+     * Returns true if the this authentication provider is no longer valid. If true, users will need to create a new
+     * DeviceClient instance to get a new authentication provider. The most common case for this is if the user
+     * provides a SAS token, but no symmetric key, and that SAS token has expired. At that point, the user's client
+     * won't be able to authenticate anymore.
+     * @return true if the this authentication provider is no longer valid. False otherwise
```

Comment:
ok, this answers a question that I asked above - why do we want to expose something like this? Wouldn't it be easier to simply state that if you create an authentication provider using a shared access signature, then you cannot renew tokens and your client will be unusable once your token expires.

Reply:
We don't expose this. These AuthenticationProvider classes are strictly internal. This particular function is used only in a few special cases where we throw an exception. I'd love to get rid of it, but that would be a breaking change.

Reply:
I am  confused here again, can't they hold the logic to provide a new SAS token in their implementation of the SasTokenProvider?

Reply:
that way they would still be able to keep using the DeviceClient.

Reply:
it's just not a great experience if I have to keep creating a device client every time my SAS expires.

Reply:
> I am confused here again, can't they hold the logic to provide a new SAS token in their implementation of the SasTokenProvider?

They can. They do not need to create a new device client each time their SAS token expires. This particular method only raises issues if the user creates a device client instance through a old constructor that allows for a single SAS token to be provided. In the case that the user uses that old constructor, and their SAS token expires, then this method returns true, signaling to the Transport layer to fire a notification to the user.

For users of the new constructor that takes the SasTokenProvider interface, they will never need to create a new device client because they will always be capable of providing new SAS tokens.